### DISCUSSION TITLE
How to get all rows from supabase db where user id is NOT in an array?

### DISCUSSION THREAD

**Discussion by nalnir** [supabase/supabase#29944] [Questions]
Im running into the issue of getting data from rewards collection where userId is NOT inside the users array. 
I asked on supabase discord channel and stack overflow but did not get an answer that works, so I decided to ask here:

I have a collection `rewards` with column `users` that can be NULL, empty array or array of uuid
I want to get all the rows that either have users column NULL, empty or do NOT contain `userId`

```
supabase
        .from('rewards')
        .select(`*`)
        .or(
          `users.not.cs{${userId}},users.is.null`,
        ),
  }),
```
  
4. Another approach I tried is:
` .or(not(users.cs.${userId$.get()}),users.is.null)`

5. I doubled checked that `userId` returns uuid and I also tried to just pass a user id directly to the call.

6. Both approaches throw an error: 
`[Error: "failed to parse logic tree ((users.not.cs{4347c7ce-8827-4da0-a8c0-3021c95f5092},users.is.null))" (line 1, column 16)]`

Any ideas what I'm missing here?

**Comment by GaryAustin1** [comment]
https://docs.postgrest.org/en/v12/references/api/tables_views.html#logical-operators

So maybe:
```
`.or(not.users.cs.{${userId$.get()}},users.is.null)`
```
Or 
```
.or(not.users.cs."[${userId$.get()}]",users.is.null)`
```

**Reply by nalnir** [reply]
I tried first one:
```
.or(`not.users.cs.{${userId$.get()}},users.is.null`),
```

And second:
```
.or(`not.users.cs.{${userId$.get()}},users.is.null`),
```

But both give error:
1st - `[Error: "failed to parse logic tree ((not.users.cs.{4347c7ce-8827-4da0-a8c0-3021c95f5092},users.is.null))" (line 1, column 8)]`
2nd - `[Error: "failed to parse logic tree ((not.users.cs."[4347c7ce-8827-4da0-a8c0-3021c95f5092]",users.is.null))" (line 1, column 8)]`

**Comment by nalnir** [comment]
Any other ides why this is not working for me?

**Reply by GaryAustin1** [reply]
I just ran this with supabase-js:
`await supabase.from('cat').select('testarray').or('testarray.not.cs.{cat},testarray.is.null')`
And it runs and gets all rows without cat in the array (both nulls and one with dog.

This is the way you say failed for you to begin with?

**Reply by nalnir** [reply]
Yes that produced the Logic tree error as well. However I've refactored my `users` column so it can be null or an array of objects that looks like this:
```
[
  {
    "id": "4347c7ce-8827-4da0-a8c0-3021c95f5092",
    "viewed": false
  }
]
```

I tweaked my call to:
```
 .or(`users.is.null, not(users.cs.'[{"id": "${userId$.get()}"}]')`)
 ```
 
 But I'm getting the same issue:
 `[Error: "failed to parse logic tree ((users.is.null,not(users.cs.'[{"id": "4347c7ce-8827-4da0-a8c0-3021c95f5092"}]')))"`
 
 I also tried this:
 ```
 .or(`users.is.null, users.not.cs.'[{"id": "${userId$.get()}"}]'`),
 ```
 
 But I am getting: `[Error: malformed array literal: "'[{"id": "4347c7ce-8827-4da0-a8c0-3021c95f5092"}]'"]`

**Reply by GaryAustin1** [reply]
I don't know what to think as the code I showed you runs fine on the current supabase-js.
Your new stuff is now having to use JSON in the mix... I've not used that enough with filters
 to offer more suggestions.

**Reply by nalnir** [reply]
Yeah I have not idea why it's not working. I guess I'll just have to leave this for now, cause I cannot come up with anything else. Thanks for looking into it.
